In [3]:
import random
import numpy as np
import pandas as pd
import dqn
import yfinance as yf
# import drqn
import dqn_markov


random.seed(69)
TEST_RATIO = 1
# ── paths ─────────────────────────────────────────────────────────────
CSV_A = '../dataset/CORN_USD_2005_2020.csv'   # ← change to your file
CSV_B = '../dataset/WHEAT_USD_2005_2020.csv'   # ← change to your file
CSV_C = '../data-scripts/data.csv'
CSV_D = '../data-scripts/data_one_regime.csv'
# ─────────────────────────────────────────────────────────────────────
df_as = []
df_bs = []
df_a_1 = pd.read_csv(CSV_A).dropna()
df_b_1 = pd.read_csv(CSV_B).dropna()
df_a_1 = df_a_1[:int(0.9*len(df_a_1))]
df_b_1 = df_b_1[:int(0.9*len(df_b_1))]


df_a_2 = pd.read_csv(CSV_C)[['s1']].dropna()
df_b_2 = pd.read_csv(CSV_C)[['s2']].dropna()
df_a_2 = df_a_2.rename(columns={"s1": "close"})
df_b_2 = df_b_2.rename(columns={"s2": "close"})

df_a_3 = pd.read_csv(CSV_D)[['s1']].dropna()
df_b_3 = pd.read_csv(CSV_D)[['s2']].dropna()
df_a_3 = df_a_3.rename(columns={"s1": "close"})
df_b_3 = df_b_3.rename(columns={"s2": "close"})

df_a_4 = yf.download("EOG", period="max", multi_level_index=False)[["Close"]].rename(columns={"Close": "close"})
df_b_4 = yf.download("XOM", period="max", multi_level_index=False)[["Close"]].rename(columns={"Close": "close"})
startindind = max(df_a_4.index.min(), df_b_4.index.min())
df_a_4, df_b_4 = df_a_4.loc[startindind:], df_b_4.loc[startindind:]

df_as = [df_a_1, df_a_3]
df_bs = [df_b_1, df_b_3]
for i in range(len(df_as)):
    df_as[i]=df_as[i][:int(TEST_RATIO*len(df_as[i]))]
    df_bs[i]=df_bs[i][:int(TEST_RATIO*len(df_bs[i]))]

agents = [dqn_markov, dqn]

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [4]:

iters = 15
arrs = np.zeros((3, 3, iters))
for j in range(len(agents)):
    for i in range(len(df_as)):
        for it in range(iters):
            ag, ret = agents[j].dqn(df_as[i], df_bs[i], random.randint(0, 10000000))
            arrs[j][i][it]=ret

Episode   10/4000  return=  -0.303  avg10=  -0.078  ε=0.993  mem=60
Episode   20/4000  return=  -0.314  avg10=  -0.154  ε=0.986  mem=140
Episode   30/4000  return=  +0.103  avg10=  -0.134  ε=0.979  mem=220
Episode   40/4000  return=  -0.005  avg10=  -0.193  ε=0.972  mem=280
Episode   50/4000  return=  -0.170  avg10=  -0.122  ε=0.966  mem=350
Episode   60/4000  return=  -0.166  avg10=  -0.109  ε=0.959  mem=400
Episode   70/4000  return=  -0.076  avg10=  -0.097  ε=0.952  mem=440
Episode   80/4000  return=  +0.063  avg10=  -0.007  ε=0.946  mem=500
Episode   90/4000  return=  -0.050  avg10=  -0.065  ε=0.939  mem=570
Episode  100/4000  return=  -0.350  avg10=  -0.112  ε=0.932  mem=650
Episode  110/4000  return=  -0.055  avg10=  -0.039  ε=0.926  mem=730
Episode  120/4000  return=  -0.087  avg10=  -0.061  ε=0.919  mem=800
Episode  130/4000  return=  -0.338  avg10=  -0.116  ε=0.913  mem=860
Episode  140/4000  return=  +0.142  avg10=  -0.077  ε=0.907  mem=940
Episode  150/4000  return=  -0.195 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  -0.161  avg10=  -0.060  ε=0.993  mem=50
Episode   20/4000  return=  -0.095  avg10=  -0.083  ε=0.986  mem=100
Episode   30/4000  return=  -0.054  avg10=  -0.055  ε=0.979  mem=130
Episode   40/4000  return=  -0.186  avg10=  -0.086  ε=0.972  mem=180
Episode   50/4000  return=  -0.083  avg10=  -0.147  ε=0.966  mem=210
Episode   60/4000  return=  -0.116  avg10=  -0.049  ε=0.959  mem=280
Episode   70/4000  return=  -0.108  avg10=  -0.110  ε=0.952  mem=320
Episode   80/4000  return=  -0.043  avg10=  -0.023  ε=0.946  mem=390
Episode   90/4000  return=  -0.090  avg10=  -0.127  ε=0.939  mem=430
Episode  100/4000  return=  -0.129  avg10=  -0.066  ε=0.932  mem=470
Episode  110/4000  return=  -0.090  avg10=  -0.100  ε=0.926  mem=510
Episode  120/4000  return=  -0.141  avg10=  -0.075  ε=0.919  mem=560
Episode  130/4000  return=  -0.168  avg10=  -0.139  ε=0.913  mem=601
Episode  140/4000  return=  -0.064  avg10=  -0.118  ε=0.907  mem=631
Episode  150/4000  return=  -0.031 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  -0.120  avg10=  -0.087  ε=0.993  mem=60
Episode   20/4000  return=  -0.053  avg10=  -0.104  ε=0.986  mem=110
Episode   30/4000  return=  -0.126  avg10=  -0.117  ε=0.979  mem=170
Episode   40/4000  return=  -0.048  avg10=  -0.173  ε=0.972  mem=210
Episode   50/4000  return=  -0.112  avg10=  -0.120  ε=0.966  mem=260
Episode   60/4000  return=  -0.366  avg10=  -0.140  ε=0.959  mem=290
Episode   70/4000  return=  -0.146  avg10=  -0.114  ε=0.952  mem=360
Episode   80/4000  return=  -0.028  avg10=  -0.078  ε=0.946  mem=410
Episode   90/4000  return=  +0.294  avg10=  -0.006  ε=0.939  mem=450
Episode  100/4000  return=  -0.191  avg10=  -0.026  ε=0.932  mem=490
Episode  110/4000  return=  -0.143  avg10=  -0.105  ε=0.926  mem=530
Episode  120/4000  return=  -0.015  avg10=  -0.064  ε=0.919  mem=620
Episode  130/4000  return=  -0.075  avg10=  -0.145  ε=0.913  mem=650
Episode  140/4000  return=  -0.158  avg10=  -0.092  ε=0.907  mem=683
Episode  150/4000  return=  -1.058 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  -0.037  avg10=  -0.095  ε=0.993  mem=60
Episode   20/4000  return=  -0.042  avg10=  -0.071  ε=0.986  mem=110
Episode   30/4000  return=  -0.090  avg10=  +0.030  ε=0.979  mem=140
Episode   40/4000  return=  -0.178  avg10=  -0.126  ε=0.972  mem=190
Episode   50/4000  return=  +0.585  avg10=  -0.129  ε=0.966  mem=230
Episode   60/4000  return=  +0.026  avg10=  -0.079  ε=0.959  mem=280
Episode   70/4000  return=  -0.017  avg10=  -0.116  ε=0.952  mem=340
Episode   80/4000  return=  -0.128  avg10=  -0.109  ε=0.946  mem=400
Episode   90/4000  return=  +0.062  avg10=  -0.013  ε=0.939  mem=460
Episode  100/4000  return=  -0.261  avg10=  -0.077  ε=0.932  mem=490
Episode  110/4000  return=  +0.055  avg10=  -0.020  ε=0.926  mem=530
Episode  120/4000  return=  +0.008  avg10=  -0.061  ε=0.919  mem=580
Episode  130/4000  return=  -0.025  avg10=  -0.058  ε=0.913  mem=640
Episode  140/4000  return=  -0.140  avg10=  -0.073  ε=0.907  mem=690
Episode  150/4000  return=  +0.047 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  +0.039  avg10=  -0.144  ε=0.993  mem=80
Episode   20/4000  return=  -0.184  avg10=  -0.078  ε=0.986  mem=110
Episode   30/4000  return=  +0.991  avg10=  -0.009  ε=0.979  mem=180
Episode   40/4000  return=  -0.296  avg10=  -0.107  ε=0.972  mem=230
Episode   50/4000  return=  -0.165  avg10=  +0.022  ε=0.966  mem=270
Episode   60/4000  return=  -0.095  avg10=  -0.097  ε=0.959  mem=310
Episode   70/4000  return=  +0.348  avg10=  -0.049  ε=0.952  mem=360
Episode   80/4000  return=  -0.091  avg10=  -0.061  ε=0.946  mem=410
Episode   90/4000  return=  -0.082  avg10=  -0.092  ε=0.939  mem=440
Episode  100/4000  return=  -0.125  avg10=  -0.081  ε=0.932  mem=520
Episode  110/4000  return=  -0.125  avg10=  -0.120  ε=0.926  mem=570
Episode  120/4000  return=  +0.014  avg10=  -0.056  ε=0.919  mem=630
Episode  130/4000  return=  -0.159  avg10=  -0.084  ε=0.913  mem=690
Episode  140/4000  return=  -0.096  avg10=  -0.012  ε=0.907  mem=760
Episode  150/4000  return=  -0.272 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  -0.150  avg10=  -0.147  ε=0.993  mem=80
Episode   20/4000  return=  -0.071  avg10=  -0.075  ε=0.986  mem=140
Episode   30/4000  return=  +0.082  avg10=  -0.096  ε=0.979  mem=220
Episode   40/4000  return=  +0.007  avg10=  -0.131  ε=0.972  mem=250
Episode   50/4000  return=  -0.083  avg10=  -0.110  ε=0.966  mem=300
Episode   60/4000  return=  -0.068  avg10=  -0.182  ε=0.959  mem=360
Episode   70/4000  return=  -0.012  avg10=  -0.109  ε=0.952  mem=380
Episode   80/4000  return=  +0.041  avg10=  -0.174  ε=0.946  mem=430
Episode   90/4000  return=  +0.038  avg10=  -0.076  ε=0.939  mem=510
Episode  100/4000  return=  +0.026  avg10=  -0.093  ε=0.932  mem=571
Episode  110/4000  return=  +0.023  avg10=  -0.081  ε=0.926  mem=631
Episode  120/4000  return=  -0.045  avg10=  -0.054  ε=0.919  mem=711
Episode  130/4000  return=  -0.157  avg10=  -0.129  ε=0.913  mem=781
Episode  140/4000  return=  -0.193  avg10=  -0.168  ε=0.907  mem=831
Episode  150/4000  return=  -0.133 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  -0.140  avg10=  -0.096  ε=0.993  mem=70
Episode   20/4000  return=  -0.188  avg10=  -0.099  ε=0.986  mem=130
Episode   30/4000  return=  -0.074  avg10=  -0.080  ε=0.979  mem=170
Episode   40/4000  return=  -0.213  avg10=  -0.060  ε=0.972  mem=230
Episode   50/4000  return=  -0.080  avg10=  -0.146  ε=0.966  mem=290
Episode   60/4000  return=  -0.150  avg10=  -0.145  ε=0.959  mem=350
Episode   70/4000  return=  -0.022  avg10=  -0.124  ε=0.952  mem=410
Episode   80/4000  return=  +0.003  avg10=  -0.074  ε=0.946  mem=440
Episode   90/4000  return=  -0.065  avg10=  -0.077  ε=0.939  mem=500
Episode  100/4000  return=  -0.075  avg10=  -0.171  ε=0.932  mem=560
Episode  110/4000  return=  -0.053  avg10=  -0.086  ε=0.926  mem=600
Episode  120/4000  return=  -0.107  avg10=  -0.149  ε=0.919  mem=660
Episode  130/4000  return=  -0.115  avg10=  -0.052  ε=0.913  mem=700
Episode  140/4000  return=  -0.014  avg10=  -0.028  ε=0.907  mem=730
Episode  150/4000  return=  -0.061 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  -0.129  avg10=  -0.120  ε=0.993  mem=90
Episode   20/4000  return=  -0.068  avg10=  -0.115  ε=0.986  mem=150
Episode   30/4000  return=  -0.221  avg10=  -0.062  ε=0.979  mem=210
Episode   40/4000  return=  -0.423  avg10=  -0.045  ε=0.972  mem=240
Episode   50/4000  return=  +0.171  avg10=  -0.105  ε=0.966  mem=290
Episode   60/4000  return=  -0.025  avg10=  -0.082  ε=0.959  mem=360
Episode   70/4000  return=  -0.196  avg10=  -0.240  ε=0.952  mem=410
Episode   80/4000  return=  -0.167  avg10=  -0.142  ε=0.946  mem=460
Episode   90/4000  return=  -0.153  avg10=  -0.115  ε=0.939  mem=530
Episode  100/4000  return=  -0.140  avg10=  -0.174  ε=0.932  mem=590
Episode  110/4000  return=  -0.085  avg10=  -0.096  ε=0.926  mem=660
Episode  120/4000  return=  +0.424  avg10=  -0.088  ε=0.919  mem=700
Episode  130/4000  return=  -0.128  avg10=  -0.101  ε=0.913  mem=750
Episode  140/4000  return=  -0.114  avg10=  -0.101  ε=0.907  mem=819
Episode  150/4000  return=  -0.121 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  -0.161  avg10=  +0.016  ε=0.993  mem=50
Episode   20/4000  return=  -0.171  avg10=  -0.093  ε=0.986  mem=120
Episode   30/4000  return=  +0.059  avg10=  -0.064  ε=0.979  mem=153
Episode   40/4000  return=  -0.038  avg10=  -0.077  ε=0.972  mem=183
Episode   50/4000  return=  -0.114  avg10=  -0.140  ε=0.966  mem=203
Episode   60/4000  return=  -0.052  avg10=  -0.060  ε=0.959  mem=243
Episode   70/4000  return=  -0.192  avg10=  -0.131  ε=0.952  mem=283
Episode   80/4000  return=  -0.014  avg10=  -0.097  ε=0.946  mem=333
Episode   90/4000  return=  -0.301  avg10=  -0.134  ε=0.939  mem=363
Episode  100/4000  return=  -0.094  avg10=  +0.004  ε=0.932  mem=393
Episode  110/4000  return=  -0.012  avg10=  -0.049  ε=0.926  mem=433
Episode  120/4000  return=  -0.334  avg10=  -0.015  ε=0.919  mem=463
Episode  130/4000  return=  -0.158  avg10=  -0.142  ε=0.913  mem=523
Episode  140/4000  return=  -0.159  avg10=  -0.050  ε=0.907  mem=563
Episode  150/4000  return=  +0.217 

c:\Users\lpset\Desktop\code\Uni\RP\trading-final\dqn_markov.py:62: RuntimeWarning: divide by zero encountered in log
  log_alpha = np.log(model.startprob_) + log_emit[0]


Episode   10/4000  return=  +0.066  avg10=  -0.180  ε=0.993  mem=40
Episode   20/4000  return=  +0.025  avg10=  -0.076  ε=0.986  mem=70
Episode   30/4000  return=  -0.122  avg10=  -0.081  ε=0.979  mem=100
Episode   40/4000  return=  -0.071  avg10=  -0.083  ε=0.972  mem=140
Episode   50/4000  return=  +0.189  avg10=  -0.107  ε=0.966  mem=220
Episode   60/4000  return=  -0.008  avg10=  -0.060  ε=0.959  mem=270
Episode   70/4000  return=  -0.126  avg10=  -0.092  ε=0.952  mem=310
Episode   80/4000  return=  -0.004  avg10=  +0.007  ε=0.946  mem=360
Episode   90/4000  return=  -0.128  avg10=  -0.117  ε=0.939  mem=410
Episode  100/4000  return=  +0.013  avg10=  -0.179  ε=0.932  mem=460
Episode  110/4000  return=  -0.115  avg10=  -0.156  ε=0.926  mem=500
Episode  120/4000  return=  -0.160  avg10=  -0.093  ε=0.919  mem=562
Episode  130/4000  return=  -0.173  avg10=  -0.056  ε=0.913  mem=602
Episode  140/4000  return=  -0.227  avg10=  -0.169  ε=0.907  mem=622
Episode  150/4000  return=  +0.042  

In [5]:
for j in range(len(agents)):
    for i in range(len(df_as)):
        print("Agent number " + str(j) + " on data number " + str(i) + " achieved mean " + str(np.mean(arrs[j][i])) + " and sd " + str(np.std(arrs[j][i])))

Agent number 0 on data number 0 achieved mean 5.338629626134315 and sd 12.312637062026162
Agent number 0 on data number 1 achieved mean 371.5716658276296 and sd 419.4618886432791
Agent number 1 on data number 0 achieved mean 8.590900990234584 and sd 11.251559168099613
Agent number 1 on data number 1 achieved mean 768.0775389573689 and sd 1071.2238168806764


Results so far:

Markov DQN: 5.3 +- 12.3 on real data

Markov DQN 371.6 +- 419.4 on generated data

DQN: 8.6 +- 11.2 on real data

DQN: 768.1 +- 1071.2 on generated data